### PR TITLE
Fix issue with single input cat

### LINKED ISSUE #97695: [indcutor] AssertionError isinstance(layout.view, ir.ReinterpretView)
### 🐛 Describe the bug

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire 

~~~
Traceback (most recent call last):
  File "/scratch/anijain/work/pytorch/failures/nn.py", line 43, in <module>
    compiled = compile_fx_inner(mod, args)
  File "/scratch/anijain/work/pytorch/torch/_dynamo/debug_utils.py", line 622, in debug_wrapper
    compiled_fn = compiler_fn(gm, example_inputs)
  File "/scratch/anijain/work/pytorch/torch/_inductor/debug.py", line 239, in inner
    return fn(*args, **kwargs)
  File "/scratch/anijain/work/env/lib/python3.9/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/scratch/anijain/work/pytorch/torch/_inductor/compile_fx.py", line 184, in compile_fx_inner
    compiled_fn = graph.compile_to_fn()
  File "/scratch/anijain/work/pytorch/torch/_inductor/graph.py", line 652, in compile_to_fn
    return self.compile_to_module().call
  File "/scratch/anijain/work/pytorch/torch/_dynamo/utils.py", line 166, in time_wrapper
    r = func(*args, **kwargs)
  File "/scratch/anijain/work/pytorch/torch/_inductor/graph.py", line 624, in compile_to_module
    code, linemap = self.codegen()
  File "/scratch/anijain/work/pytorch/torch/_inductor/graph.py", line 575, in codegen
    self.scheduler.codegen()
  File "/scratch/anijain/work/pytorch/torch/_dynamo/utils.py", line 166, in time_wrapper
    r = func(*args, **kwargs)
  File "/scratch/anijain/work/pytorch/torch/_inductor/scheduler.py", line 1216, in codegen
    self.codegen_extern_call(node)
  File "/scratch/anijain/work/pytorch/torch/_inductor/scheduler.py", line 1137, in codegen_extern_call
    scheduler_node.allocate()
  File "/scratch/anijain/work/pytorch/torch/_inductor/scheduler.py", line 273, in allocate
    V.graph.wrapper_code.codegen_allocation(self.node)
  File "/scratch/anijain/work/pytorch/torch/_inductor/codegen/wrapper.py", line 597, in codegen_allocation
    assert isinstance(layout.view, ir.ReinterpretView)
AssertionError
~~~

### Error logs

_No response_

### Minified repro

~~~
import torch._inductor.overrides

import torch
from torch import tensor, device
import torch.fx as fx
from torch._dynamo.testing import rand_strided
from math import inf
from torch.fx.experimental.proxy_tensor import make_fx

import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config

# REPLACEABLE COMMENT FOR TESTING PURPOSES

# torch version: 2.1.0a0+fb
# torch cuda version: 11.4.0
# CUDA Info:
# nvcc not found
# GPU Hardware Info:
# NVIDIA PG509-210 : 8


from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()



    def forward(self, arg3_1, relu, permute_1):
        addmm_1 = torch.ops.aten.addmm.default(arg3_1, relu, permute_1);  arg3_1 = relu = permute_1 = None
        cat_2 = torch.ops.aten.cat.default([addmm_1], 1);  addmm_1 = None
        return (cat_2,)

args = [((96,), (1,), torch.float32, 'cuda'), ((10, 256), (256, 1), torch.float32, 'cuda'), ((256, 96), (1, 256), torch.float32, 'cuda')]
args = [rand_strided(sh, st, dt, dev) for (sh, st, dt, dev) in args]
mod = make_fx(Repro(), tracing_mode='real')(*args)

from torch._inductor.compile_fx import compile_fx_inner
from torch._dynamo.debug_utils import same_two_models

compiled = compile_fx_inner(mod, args)
ref = compiled(args)
torch.cuda.synchronize() # Ensures that segfaults are surfaced
~~~

### Versions

NA

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97822

Fixes #97695

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by ngimel — APPROVED** [review]

File: torch/_inductor/decomposition.py [hunk 1]
```diff
@@ -371,6 +371,13 @@ def baddbmm(self, batch1, batch2, beta=1, alpha=1):
     return self + result
 
 
+@register_decomposition([aten.cat.default])
+def cat(tensors, dim=0):
+    if len(tensors) == 1:
```
Comment (ngimel) [inline torch/_inductor/decomposition.py:376]: we already have this special case in `cat` lowering, why do we need it in decomposition?

Comment (jansel) [reply torch/_inductor/decomposition.py:376]: The issue fixed in this PR is `pattern_matcher.py` is matching a mm+cat fusion and that fused pattern didn't handle this case.  So it wasn't doing through the normal lowering.

Comment (ngimel) [reply torch/_inductor/decomposition.py:376]: But I've just tried and original issue is fixed just by changes in `patter_matcher.py`, this change is not required, `cat_tuned_op` is used in lowering patterns only, they never go through decompositions I thought?

Comment (jansel) [reply torch/_inductor/decomposition.py:376]: Correct.  Though matching that pattern can prevent other patterns from getting matched (for example an mm to addmm), so it is good to remove the pointless cat from the graph.

**Review by anijain2305 — APPROVED** [review]

**Pull request closed without merging**